lib/pengine/rules.c

```diff
@@ -1043,9 +1047,24 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
         if (safe_str_eq(type, "string")) {
             cmp = strcasecmp(h_val, value);
 
+        } else if (safe_str_eq(type, "integer")) {
+            long long h_val_f = crm_parse_ll(h_val, NULL);
+            long long value_f = crm_parse_ll(value, NULL);
+
+            if (h_val_f < value_f) {
+                cmp = -1;
+            } else if (h_val_f > value_f) {
+                cmp = 1;
+            } else {
+                cmp = 0;
+            }
+
         } else if (safe_str_eq(type, "number")) {
-            int h_val_f = crm_parse_int(h_val, NULL);
-            int value_f = crm_parse_int(value, NULL);
+            double h_val_f;
+            double value_f;
+
+            pcmk__scan_double(h_val, &h_val_f, NULL, NULL);
+            pcmk__scan_double(value, &value_f, NULL, NULL);
```

Comment:
We may want to check for parse errors here. I didn't do so in the first pass because:
- I haven't given much thought to what we should do if one or both values fail to parse.
- We haven't done any parse error checking in this function prior to this PR, which is already large.

I'm all ears if you have an opinion. I think the simplest approach would be to return false immediately without doing any further processing. Otherwise, there's no **obviously correct** approach for deciding how to handle how comparisons should be handled (e.g., is -3 less than a parse failure?).

Reply:
Checking for parse errors could be considered a backwards compatibility break. On the other hand, you could make the argument that parse failure behavior has been undefined in the past, since the values are expected to be numbers if type="number" is used.

In practice, a parse failure has always yielded the value -1.

Reply:
Note to self: Update/add regression tests if we check for parse errors.

Reply:
Now you see what I meant about using return codes to force developers to think about error handling : )

I'm thinking a parse error should switch the comparison type to string. At least then we're returning a value that has some meaning (always returning false still assigns an order to parse errors, but that order varies by operation).

Reply:
Good idea, and yep.

Reply:
> always returning false still assigns an order to parse errors

Only in a sense. It would cause the expression to evaluate to false, regardless of the operation. The comparison isn't in a separate function. The comparison and the handling of its result are both in pe__eval_attr_expr.